linear torsion angle fix

davidfarinajr opened the pull request:
I added a method within `get_torsions` that checks for linear (or near-linear within a tolerance) angles in a torsion.  If an angle is near 180 degrees, we look farther down the molecule for an atom to define the torsion.  I also added `angles`, `description`, and `center_atoms` attribute to torsion class.

davidfarinajr commented:
<img width="415" alt="Screen Shot 2020-03-16 at 9 04 04 PM" src="https://user-images.githubusercontent.com/32377555/76811947-b91d1600-67c9-11ea-966d-bc3780939189.png">
It's helpful for molecules like FCC#CC.  Previously, the get_torsions method would return 2 torsions.  With this PR, 1 torsion is returned where the terminal carbons are the pivot atoms.

ndharms commented on autotst/geometry.py:
```diff
@@ -102,13 +102,20 @@ def __init__(
         self.atom_indices = atom_indices
         self.dihedral = dihedral
         self.mask = mask
+        self.angles = angles
```
If a torsion is made up of atoms i, j, k, and l, `self.angles` are the angles between atoms i, j, and k and atoms j, k, and l. Correct? This feels like redundant information considering `Angle` objects should exist with this information.

davidfarinajr replied:
I agree, we don't really need this.  It's also the reason why the unit tests failed.  I removed it.

ndharms commented on autotst/geometry.py:
```diff
@@ -102,13 +102,20 @@ def __init__(
         self.mask = mask
+        self.angles = angles
+        self.description = description.lower()
+        self.center_atoms = center_atoms
```
This feels redundant because `center_atoms` are given in `atom_indices` as the 2nd and 3rd element of a 4 element list.

davidfarinajr replied:
`center_atoms` is a list of atoms between the 2 pivot atoms.  In most cases, it will be an empty list.  However, If you have a torsion with a rotation axis that spans more than two atoms (like in H3C-C#C-CH3), the center atoms are the atoms between the 2 pivots (the 2 triply bonded carbons in this case).

ndharms replied:
Oh, I'm understanding this. This makes a lot more sense, maybe add a comment about this if that's the case then.

ndharms commented on autotst/geometry.py:
```diff
@@ -102,13 +102,20 @@ def __init__(
+        self.center_atoms = center_atoms
         self.reaction_center = reaction_center
-
+        self.center_atoms = center_atoms
```
Repeated

ndharms commented on autotst/species.py:
```diff
@@ -433,12 +433,34 @@ def get_angles(self):
+            Checks the two angles in the torsion to  
+            """
+            i,j,k,l = tor
+            angles = ((0,self.ase_molecule.get_angle(i, j, k)), (-1,self.ase_molecule.get_angle(j, k, l)))
```
Rather than getting the angles again, we may want to grab the angles from the `Angle` objects that are attached to a conformer already. This may require us to call `get_angles` before (or while) we check the angles.

davidfarinajr replied:
`get_angles` returns a list of all angles, so it's a bit difficult to find the angles of interest

ndharms commented on autotst/species.py:
```diff
@@ -502,6 +519,44 @@ def get_torsions(self):
+                endpoints = check_angles(torsion_tup)
+                center_atoms = []
+   
+                while len(endpoints) > 0:
```
Please add comments here, it's hard to understand what's going on here.

ndharms replied:
I'm not sure if I'm tired cuz it's late, but this is a bit confusing. Could you explain this section of the code?

davidfarinajr replied:
I added some comments, although it is very confusing.  Maybe we could try to think of a cleaner algorithm?

ndharms commented on autotst/species.py:
```diff
@@ -502,6 +519,44 @@ def get_torsions(self):
+                            vertex = torsion_tup[1]
+                        else:
+                            vertex = torsion_tup[2]
+                        ID = torsion_tup[index]
```
Maybe rename ID as atom_id, I'm personally not a big fan of all caps variables unless they are constants. Here's a decent [style guide](https://realpython.com/python-pep8/)

ndharms commented:
Also, testing failed and coverage is down, these will need to be addressed before we can merge this in.

davidfarinajr commented:
<img width="383" alt="Screen Shot 2020-03-17 at 10 34 51 AM" src="https://user-images.githubusercontent.com/32377555/76867251-a8a58380-683b-11ea-862e-46479a666dbe.png">
<img width="619" alt="Screen Shot 2020-03-17 at 10 40 39 AM" src="https://user-images.githubusercontent.com/32377555/76867390-d7235e80-683b-11ea-9895-fdd17d02a64e.png">

This is rotor scan for SMILES `FCC#CCF` with `C1` and `C2` as the pivot atoms (`F1-C1-C2-F2`).  Currently, AutoTST generates 2 torsions for this molecule: `(F1-C1-C3-C4)` and `(F2-C2-C4-C3)`.  However, both of these scans failed because `C1-C3-C4` and `C2-C4-C3` are linear.  The goal of this PR is to identify rotors where the axis of rotation has more than 2 atoms (pivot atoms are not directly connected in graph).

The `center_atoms` are the atoms in between the pivot atoms (in this case `C3` and `C4`).

ndharms commented on autotst/species_test.py:
```diff
@@ -67,10 +68,11 @@ def test_get_angles(self):
+            torsions = conformer.get_torsions()
+            self.assertIsInstance(torsions,list)
+            self.assertIsInstance(torsions[0],Torsion)
+            self.assertEquals(len(torsions),1)
```
Good test to check that we're finding the right number of torsions! Can you also add one to check the number of center atoms for the C-C#C-C case?

davidfarinajr commented:
@nateharms, I added `center_atoms` test.  I also realized we need to consider the `center_atoms` when getting the mask, so I modified the `get_mask` method and refactored the `get_torsions` method a bit.  Can you take a look at the new changes?

ndharms commented:
Will do, I'll wait a sec for the checks to pass, then I'll review and merge if it's all good.
